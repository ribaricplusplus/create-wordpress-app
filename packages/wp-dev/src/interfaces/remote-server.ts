export default interface RemoteServer {
	isConnected: boolean;

	connect: ( config: any ) => Promise< void >;
	disconnect: () => Promise< void >;

	execScript: (
		script: string,
		options?: ExecScriptOptions
	) => Promise< { output: string; code: number } >;

	download: ( source: string, destination: string ) => Promise< void >;
}

interface ExecScriptOptions {
	logOutput?: boolean;
}
