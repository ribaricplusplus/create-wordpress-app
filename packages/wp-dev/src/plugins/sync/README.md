# Sync plugin

## Usage notes

- Uses SSH with public / private keys to connect to the remote server. Expects that this is already configured properly.
