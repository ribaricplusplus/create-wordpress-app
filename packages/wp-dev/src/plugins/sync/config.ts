import path from 'path';
import os from 'os';
import Joi from 'joi';

import SshConfig from './interfaces/ssh-config';
import { WpDevConfiguration } from '../../types';
import Config from './interfaces/config';

export function extendConfig( config: WpDevConfiguration ): Config {
	const defaults: {
		ssh: Partial< SshConfig >;
		wordPressPath: string;
	} = {
		wordPressPath: '/var/www/html',
		ssh: {
			privateKeyPath: path.join( os.homedir(), '.ssh/id_rsa' ),
			port: 22,
		},
	};

	const extendedConfig = {
		...defaults,
		...config,
		ssh: {
			...defaults.ssh,
			...config?.ssh,
		},
	};

	const schema = Joi.object( {
		wordPressPath: Joi.string().min( 1 ),
		ssh: Joi.object( {
			host: Joi.string().min( 1 ),
			port: Joi.number().min( 1 ).integer(),
			username: Joi.string().min( 1 ),
			privateKeyPath: Joi.string().min( 1 ),
		} ),
	} ).unknown( true );

	Joi.assert( extendedConfig, schema );

	return extendedConfig as Config;
}
