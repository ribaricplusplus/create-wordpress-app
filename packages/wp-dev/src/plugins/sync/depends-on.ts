export const dependsOn = {
	LocalServer: Symbol.for( 'LocalServer' ),
	RemoteServer: Symbol.for( 'RemoteServer' )
};
