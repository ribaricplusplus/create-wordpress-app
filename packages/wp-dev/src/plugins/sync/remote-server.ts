import 'reflect-metadata';
import { injectable } from 'inversify';
import { Client } from 'ssh2';
import { readFile } from 'fs/promises';
import { nanoid } from 'nanoid';
import type { SFTP } from 'ssh2';

import Config from './interfaces/config';
import IRemoteServer from '../../interfaces/remote-server';

@injectable()
export default class RemoteServer implements IRemoteServer {
	public client: Client;

	public isConnected: boolean = false;

	constructor() {
		this.client = new Client();

		this.client.on( 'error', ( err ) => {
			console.error( 'SSH client error occurred.' );
			console.error( err );
		} );

		this.client.on( 'end', () => {
			this.isConnected = false;
		} );
	}

	/**
	 * @throws If connection fails.
	 */
	async connect( { ssh }: Config ): Promise< void > {
		const privateKey = await readFile( ssh.privateKeyPath, {
			encoding: 'utf8',
		} );

		this.client.connect( {
			host: ssh.host,
			port: ssh.port,
			username: ssh.username,
			privateKey,
		} );

		return new Promise( ( resolve, reject ) => {
			this.client.once( 'ready', () => {
				this.isConnected = true;
				resolve();
			} );
		} );
	}

	async disconnect() {
		if ( ! this.isConnected || ! this.client ) {
			return;
		}

		this.client.end();

		return new Promise< void >( ( resolve, reject ) => {
			this.client.once( 'end', () => {
				this.isConnected = false;
				resolve();
			} );
		} );
	}

	async execScript( script, _options ) {
		const defaultOptions = {
			logOutput: false,
		};

		const options = {
			...defaultOptions,
			..._options,
		};

		if ( ! this.isConnected || ! this.client ) {
			throw new Error(
				'Cannot execute scripts when a connection is not established.'
			);
		}

		const sftp = await this.getSftp();

		// Write file somewhere in remote server /tmp directory
		const remoteScriptName = `create-wordpress-app-script-${ nanoid() }.sh`;
		const remoteScriptPath = `/tmp/${ remoteScriptName }`;
		let result;

		try {
			const writeStream = sftp.createWriteStream( remoteScriptPath, {
				mode: 0o776,
				encoding: 'utf8',
			} );
			writeStream.end( script );
			await new Promise( ( resolve, reject ) => {
				writeStream.once( 'error', reject );
				writeStream.once( 'finish', resolve );
			} );
			result = await new Promise( ( resolve, reject ) => {
				this.client.exec( remoteScriptPath, ( err, stream ) => {
					if ( err ) {
						reject( err );
						return;
					}

					let allOutput = '';
					const handleData = ( data ) => {
						if ( options.logOutput ) {
							console.log( data.toString() );
						}
						allOutput += data.toString();
					};
					stream.stdout.on( 'data', handleData );
					stream.stderr.on( 'data', handleData );
					stream.on( 'close', ( code ) => {
						resolve( { output: allOutput, code } );
					} );
					stream.on( 'error', reject );
				} );
			} );
		} finally {
			await new Promise< void >( ( resolve, reject ) => {
				sftp.unlink( remoteScriptPath, ( err ) => {
					if ( err ) {
						reject( err );
						return;
					}
					resolve();
				} );
			} );
		}

		return result;
	}

	async download( source, destination ) {
		const sftp = await this.getSftp();
		await new Promise< void >( ( resolve, reject ) => {
			sftp.fastGet( source, destination, ( err ) => {
				if ( err ) {
					reject( err );
					return;
				}
				resolve();
			} );
		} );
	}

	private getSftp(): Promise< SFTP > {
		if ( ! this.isConnected ) {
			throw new Error(
				'Cannot get SFTP client when not connected to a remote server.'
			);
		}

		return new Promise( ( resolve, reject ) => {
			this.client.sftp( ( err, sftp ) => {
				if ( err ) {
					reject( err );
					return;
				}

				resolve( sftp );
			} );
		} );
	}
}
