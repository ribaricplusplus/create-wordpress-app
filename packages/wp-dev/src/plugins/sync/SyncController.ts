import 'reflect-metadata';
import { injectable, inject } from 'inversify';
import { applyFilters } from '@wordpress/hooks';
import { execSync } from 'child_process'

import LocalServer from '../../interfaces/local-server';
import RemoteServer from '../../interfaces/remote-server';
import { dependsOn } from './depends-on';
import { WpDevConfiguration, CommandCommanderOptions } from '../../types';

@injectable()
export default class SyncController {
	public localServer: LocalServer;
	public remoteServer: RemoteServer

	constructor( @inject( dependsOn.LocalServer ) localServer: LocalServer, @inject( dependsOn.RemoteServer ) remoteServer: RemoteServer ) {
		this.localServer = localServer;
		this.remoteServer = remoteServer
	}

	async run( config: WpDevConfiguration, options: CommandCommanderOptions ) {
		await this.remoteServer.connect( config.ssh ?? {} )
		const missingDependencies = await this.getMissingSystemAndServerDependencies();

		if ( missingDependencies.length > 0 ) {
			for( const dependency of missingDependencies ) {
				console.error( `Error: Missing ${ dependency.type } dependency: ${ dependency.name }` );
			}
			return;
		};

		if ( ! ( await this.localServer.isStarted() ) ) {
			// We're not starting the server because we don't know what
			// parameters a user would want to start a server with... Maybe they
			// want it to start with --xdebug, for example.
			console.error( 'Local server not started. Please start it first so that synchronization can complete.' )
			return;
		}

		const downloadFilesFromServer = applyFilters( 'wp-dev-plugin-sync-download-callback', this.downloadFilesFromServer.bind ) as Function;
		await downloadFilesFromServer();

		await this.importFiles();
		await this.importDatabase();
	}

	async getMissingSystemAndServerDependencies(): Promise<Array<{ type: 'system' | 'server', name: string}>>  {

	}

	async downloadFilesFromServer() {
		const remoteExecCb = () => {
			// TODO
			execSync( 'ssh' )
			// Check whether someone else is currently doing a sync... Only 1 sync allowed at a time.
		}
		// TODO: Connect through ssh and proceed
	}

	async importFiles() {
		// TODO import files that were downloaded from server
	}

	async importDatabase() {
		// TODO import database that was download from server
	}
}
