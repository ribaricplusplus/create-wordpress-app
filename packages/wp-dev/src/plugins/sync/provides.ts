import SyncController from './SyncController';
import RemoteServer from './remote-server';
import { ServiceDefinition, ServiceRegistrationCallback } from '../../types';

type Provides = {
	SyncController: ServiceDefinition;
	RemoteServer: ServiceRegistrationCallback;
};

export const provides: Provides = {
	SyncController: {
		symbol: Symbol.for( 'SyncController' ),
		implementation: SyncController,
	},

	RemoteServer: ( container ) => {
		container
			.bind( Symbol.for( 'RemoteServer' ) )
			.to( RemoteServer )
			.inTransientScope();
	},
};
