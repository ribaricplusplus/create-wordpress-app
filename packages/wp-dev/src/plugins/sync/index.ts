import { addFilter } from '@wordpress/hooks';
import os from 'os';
import path from 'path';

import {
	CommandConfiguration,
	CommanderActionGenerator,
	PluginInitOptions,
	CommandCommanderOptions,
	WpDevConfiguration,
} from '../../types';
import { provides } from './provides';
import type SyncController from './SyncController';

const sync: CommanderActionGenerator = ( config, container ) => async (
	options: CommandCommanderOptions
) => {
	const controller = container.get(
		provides.SyncController.symbol
	) as InstanceType< typeof SyncController >;
	controller.run( config, options );
};

const commandConfig: CommandConfiguration = {
	command: [ 'sync' ],
	description: [ 'Synchronize local server with upstream.' ],
	action: sync,
};

const init = () => {
	addFilter(
		'wp-dev-commands-configurations',
		'wp-dev/plugins/sync/add-commands',
		( commands ) => [ ...commands, commandConfig ]
	);
	addFilter(
		'wp-dev-service-definitions',
		'wp-dev/plugins/sync/add-service-definitions',
		( definitions ) => {
			return { ...definitions, ...provides };
		}
	);
	addFilter(
		'wp-dev-config',
		'wp-dev/plugins/sync/config',
		( config: WpDevConfiguration ) => {
			const defaults = {
				ssh: {
					privateKeyPath: path.join( os.homedir() ),
				},
			};
		}
	);
};

export const pluginConfig = {
	init,
};
