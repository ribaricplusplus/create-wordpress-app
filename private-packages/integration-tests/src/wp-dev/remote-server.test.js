const os = require( 'os' );
const path = require( 'path' );

const RemoteServer =
	require( '@ribarich/wp-dev/src/plugins/sync/remote-server' ).default;

function createRemoteServer() {
	return new RemoteServer();
}

describe( 'RemoteServer', () => {
	describe( 'connection', () => {
		const server = createRemoteServer();

		afterEach( async () => {
			await server.disconnect();
		} );

		it( 'Can connect to remote machine', async () => {
			await expect(
				server.connect( {
					ssh: {
						host: '192.168.56.125',
						username: 'vagrant',
						privateKeyPath: path.join(
							os.homedir(),
							'.ssh/id_rsa'
						),
						port: 22,
					},
				} )
			).resolves.not.toThrow();
		} );
	} );

	describe( 'execScript', () => {
		const server = createRemoteServer();

		beforeAll( async () => {
			await server.connect( {
				ssh: {
					host: '192.168.56.125',
					username: 'vagrant',
					privateKeyPath: path.join( os.homedir(), '.ssh/id_rsa' ),
					port: 22,
				},
			} );
		} );

		afterAll( async () => {
			await server.disconnect();
		} );

		it( 'Returns expected output for ls in test folder.', async () => {
			const script = `
#/usr/bin/env bash
set -e
cd /home/vagrant/testfolder
ls
`;

			const { output, code } = await server.execScript( script );
			expect( output.length ).toBeGreaterThan( 0 );
			expect( code ).toBe( 0 );
			expect( output ).toMatchSnapshot();
		} );
	} );
} );
