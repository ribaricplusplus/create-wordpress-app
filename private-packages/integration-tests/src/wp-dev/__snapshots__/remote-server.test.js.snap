// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoteServer execScript Returns expected output for ls in test folder. 1`] = `
"file.txt
"
`;
